synapse_data_warehouse/synapse_raw/tables/V2.23.2__refresh_node_snapshot.sql

```diff
@@ -0,0 +1,56 @@
+-- Configure environment
+USE ROLE sysadmin;
+USE SCHEMA {{database_name}}.synapse_raw; --noqa: JJ01,PRS,TMP,CP02
+USE WAREHOUSE compute_medium;
+
+-- Drop all records from this table
+TRUNCATE TABLE nodesnapshots;
```

Comment:
I'm ok without truncating the table, there will be slight data loss, but we can continue copying data.  By truncating the data it will take longer than an hour to run the processing which might cause this CI/CD pipeline to timeout.

Reply:
@thomasyu888 In that case, should we remove this versioned script altogether? If we don't truncate/backfill, the table will forever be missing `version_history` for those snapshots which have this field. On the other hand, the practical aspect of having the `version_history` for every node will eventually be taken care of, as you said. That's assuming that this table behaves like a snapshot table and not a change capture table, a concern I brought up with Sandhra [here](https://sagebionetworks.slack.com/archives/C05PBT2FUHF/p1732647867155019?thread_ts=1732554003.064199&cid=C05PBT2FUHF).

Reply:
One final comment then I'm ooo.

1. There are change timestamps captures within the table when there's an action like create, update or delete.
2. Every week, there is a snapshot timestamp that captures all the actions within the week or perhaps captures whatever is the current state within the data warehouse (this we need to confirm more)


To add to further question for your question to sandrah, when new data is added, it's not always backfilled with all the variables. Which means that the data for that field will exist when there is a new change to the entity, then the snapshot will capture that on a weekly basis.

The snapshot is NOT a snapshot of what's within the database (_i think_) but rather what's in the data warehouse.

Reply:
@thomasyu888

> whether or not the weekly snapshots contain all the data that's been snapshotted previously OR just the snapshots of the actions for that week.

I queried the production data. I found approximately the same number of unique, non-deleted nodes in the `nodesnapshots` table as the number of nodes which were snapshotted in the last 8 days. I also found nodes which haven't been changed recently yet were included in the recent snapshot. So my thinking is that _every_ existing (i.e. `change_type` != 'DELETE') node, regardless of when it was last changed, is included in the weekly snapshot.

What this means in the context of this PR is that the new `version_history` column will be populated for every existing node in the next weekly snapshot without us needing to truncate and re-`COPY INTO` the `nodesnapshots` table. But any nodes which were deleted before the next snapshot and since `version_history` began being collected won't have this data "backfilled" automatically. If we don't care about including version history metadata for such nodes (~3 million nodes, or ~5% of the unique nodes in the `nodesnapshots` table), then we should get rid of this versioned script.

Reply:
Alternatively, if we do want to preserve this metadata but we don't want to re-populate the entire `nodesnapshots` table, we might be able to `DELETE` any records which have a snapshot date in the last ~3 weeks and then force `COPY INTO` just those records from the stage. That's ~90 million records, or ~2% of the total records in the `nodesnapshots` table.

Reply:
@philerooski 

What is your recommendation based off of your findings?  it sounds like what we would be missing is `version_snapshot` for the `DELETE` records?  If that is the case, those would be removed anyhow from the latest tables.

Reply:
@thomasyu888

I don't like the idea of throwing away historical data because it won't impact the most up-to-date data. 90 million records is not such a huge amount of data to copy into Snowflake. I recommend deleting and copying data selectively so that our `nodesnapshots` table in Snowflake is up-to-date with the data warehouse.

Reply:
@philerooski 

I like that idea a lot! Let's go with your recommendation.  Thanks for looking into the data!  For the userprofile snapshots, one of the columns haven't been backfilled in awhile, so for that snapshot, it may actually make sense to just do a whole sail replacement, but I would also look at the production data first.

Reply:
@thomasyu888 

What's annoying about doing this in practice is that there's no perfect way to say `COPY INTO` all files from this date onward. The only filtering we can do as part of the `COPY INTO` statement is regex. Theoretically, I could invent some horrendous pattern which says something like "all files from 2024-11-* and 2024-12-* and all files for the next 100 years after that.

Instead, we might be tempted to query the last modified date of a file by `LIST`ing a stage and `COPY INTO`ing only those files, but Snowflake gives this to us as an ugly, non-ISO string like "Wed, 2 Oct 2024 00:02:40 GMT" that can't be easily cast as a datetime. Probably we would rather parse out the `snapshot_date=*` component of the S3 prefix and cast that instead.

The bigger picture problem here is that when including `COPY INTO` statements in versioned scripts, the end state of the target table is dependent on the data in the stage -- which is constantly changing. So if there was ever a need to roll back to a previous version of the database and replay these changes we wouldn't get the same result.

/rant I'm going to go implement the horrendous regex now